### DISCUSSION TITLE
Timeout pushing image for a Web API container app

### DISCUSSION THREAD

**Discussion by ciwchris** [Azure/azure-dev#3212] [Q&A] [answered]
When trying to deploy a Web API as a Azure Container App I receive a timeout error pushing the image.

> error CONTAINER1013: Failed to push to the output registry: The request was canceled due to the configured HttpClient.Timeout of 100 seconds elapsing.

If I trim the project, set PublishTrimmed to true, then I can get it to deploy most of the time. But trimming the app breaks the app, in it's current form, and [trimming isn't officially supported for this type of app](https://github.com/dotnet/aspnetcore/issues/44997#issuecomment-1310894044).

Are there other workarounds for pushing the image without timing out?

**Comment by davidfowl** [comment]
cc @baronfel

**Reply by baronfel** [reply, marked as answer]
We don't currently have a way to set this timeout with the container tooling. I've logged https://github.com/dotnet/sdk-container-builds/issues/543 to track adding some kind of way to work around this. In the interim as a workaround I'd suggest pushing to a local docker daemon and then using `docker push` to have Docker push the image to your destination registry.

**Comment by ciwchris** [comment]
Thank you for getting it routed to the right place.

Okay, thank you. This project is just a learning experiment, so yeah, for now I push the image myself.

**Comment by tmds** [comment]
> If I trim the project, set PublishTrimmed to true, then I can get it to deploy most of the time. But trimming the app breaks the app, in it's current form, and https://github.com/dotnet/aspnetcore/issues/44997#issuecomment-1310894044.

It sounds like the timeout is hit because what gets uploaded is "large", that is: `size / 100 s > bandwidth`.

> Are there other workarounds for pushing the image without timing out?

You could try to force the upload to chunked by setting this envvar `SDK_CONTAINER_DEBUG_REGISTRY_FORCE_CHUNKED_UPLOAD=true`.

The chunks will only be 64kB which might cause a slow upload.
You can increase the size by setting `SDK_CONTAINER_REGISTRY_CHUNKED_UPLOAD_SIZE_BYTES`.

**Reply by tmds** [reply]
cc @ellismg

**Reply by baronfel** [reply]
It would also help if `azd` created binlogs for these publishes - we log trace-level diagnostics about the push to the binlog, and that would help us know if our intended chunk-based upload fallback is happening or not. We _intended_ that if the single atomic push fails for any reason (including timeouts) the SDK should fallback to uploading chunks in the way that @tmds has described.
